### PR TITLE
Ready for Vue

### PR DESCRIPTION
**Pull request by ytbryan**
This PR ships with Vue, its supported libraries & configuration which I reference partially from chrisvfritz's repo (https://github.com/chrisvfritz/rails-javascript-integrations/tree/vue-webpack) 

It includes the classic <%= javascript_pack_tag 'hello_vue' %> and another <%= javascript_pack_tag 'vue_todo' %> example that demonstrates other capabilities of Vue found in chrisvfritz's vue-webpack branch.

### CODE REVIEW

**Comment by dleavitt** [comment]
Nice! I am a big fan of vue so excited to have progress on this. And gj putting all that error messaging in the generator, future generations will thank you! 

I think what's done here is quite excessive for a generator though - it installs an entire Todo app into your Rails app.
- it modifies the database, adding a model. unless the user wants a Todo model with that schema, it means they'll have to rollback or undo that migration before proceeding. It also adds an implicit dependency on active record. A vue scaffold should not be messing with the db.
- adds routes and an api endpoint, also things users will have to delete after running the generator. Doesn't add a route to actually test the vue app though, as far as I can tell
- uses url-loader. file-loader is a better default, and I think using url-loader without file-loader may cause problems for assets too big to be base64'd (try setting a low size limit on url-loader and see what happens)
- puts all assets under `app/javascript/packs`. Everything but the entry points should be under `app/javascript`.
- puts a png of the vue logo into a gem that's going to be downloaded by everyone who uses rails. Best practices mandate using an image of [Guy Fieri](https://medium.com/friendship-dot-js/i-peeked-into-my-node-modules-directory-and-you-wont-believe-what-happened-next-b89f63d21558) instead.

Suggestion:
Make this more like the react and angular generators - just enough to demo the core functionality in a way that's easy to build upon. IMO this would be more like the hello_vue.js entrypoint included here, with the addition of a single `.vue` file (or a small app that doesn't require server-side interaction.)

Additional functionality beyond the above should be removed and made into a separate app template or a demo app.

**Comment by ytbryan** [comment]
Thanks, appreciate the feedback. Will move the todo example to a separate app template.

File: lib/tasks/webpacker.rake [hunk 1]
```diff
@@ -97,6 +97,43 @@ namespace :webpacker do
...
       exec './bin/yarn add --dev typescript ts-loader && ./bin/yarn add "core-js zone.js rxjs @angular/core @angular/common @angular/compiler @angular/platform-browser @angular/platform-browser-dynamic"'
     end
+
+    desc "Install everything needed for Vue"
+    task :vue do
+      config_path = Rails.root.join('config/webpack/shared.js')
+      config = File.read(config_path)
+
+      # Module resolution https://webpack.js.org/concepts/module-resolution/
+      if config.include?("'vue$':'vue/dist/vue.common.js'")
```
Comment (zigomir) [inline lib/tasks/webpacker.rake]: I think this won't be necessary anymore with Vue v2.2.x https://github.com/vuejs/vue/releases/tag/v2.2.0

Comment (ytbryan) [inline lib/tasks/webpacker.rake]: @zigomir thanks for informing me. I guess it needs to be updated to  `'vue$': 'vue/dist/vue.esm.js'` 

Will send in a bundle PR after I read through the entire vue 2.2.x release.

**Comment by ankitsinghaniyaz** [comment]
@ytbryan Will this disable my normal views and erb helpers?

**Comment by ytbryan** [comment]
@ankitsinghaniyaz can you elaborate and if possible, show with some code?

**Comment by ankitsinghaniyaz** [comment]
@ytbryan I mean if I use webpack and vue configuration. Will I still be able to use `.html.erb` files?

**Comment by ytbryan** [comment]
@ankitsinghaniyaz Yes, webpack configuration should not affect your existing erb view files.

**Comment by ankitsinghaniyaz** [comment]
@ytbryan I created a new app, but it will be awesome if I have some guidelines how to structure the code and how to mix it with normal rails templating.

**Comment by BigChief45** [comment]
Hi @ytbryan great work.

I am having trouble setting this up with a fresh Rails 5.0.2 application.

After creating the application, I add webpacker gem and then bundle:

`gem 'webpacker', github: 'rails/webpacker'`

Then I run `rails webpacker:install` and get this:

```
      create  app/javascript
      create  app/javascript/packs/application.js
       exist  bin
      create  bin/webpack-dev-server
      create  bin/webpack-watcher
      create  bin/webpack
      create  bin/yarn
      create  config/webpack
      create  config/webpack/development.js
      create  config/webpack/production.js
      create  config/webpack/shared.js
      append  .gitignore
         run  ./bin/yarn add webpack webpack-merge path-complete-extname babel-loader babel-core babel-preset-env coffee-loader coffee-script compression-webpack-plugin rails-erb-loader glob from "."
Yarn executable was not detected in the system.
Download Yarn at https://yarnpkg.com/en/docs/install
         run  ./bin/yarn add --dev webpack-dev-server from "."
Yarn executable was not detected in the system.
Download Yarn at https://yarnpkg.com/en/docs/install
```

At this point it seems that there is something wrong. For some reason this Yarn is not in the system. Anyway, I proceeded to install Vue with `rails webpacker:install:vue` which results in:

```
Copying the Vue example to app/javascript/packs/vue
Yarn executable was not detected in the system.
Download Yarn at https://yarnpkg.com/en/docs/install
```

The Vue examples are actually copied to `app/javascripts/packs`:

```
app/javascript/packs/
├── app.vue
├── application.js
└── hello_vue.js

0 directories, 3 files
```

I proceeded to add `<%= javascript_pack_tag 'hello_vue' %>` to my `application.html.erb` layout. I get the following errors in the Rails server:

```
ActionController::RoutingError (No route matches [GET] "/packs/hello_vue.js"):
  
actionpack (5.0.2) lib/action_dispatch/middleware/debug_exceptions.rb:53:in `call'
web-console (3.4.0) lib/web_console/middleware.rb:135:in `call_app'
web-console (3.4.0) lib/web_console/middleware.rb:20:in `block in call'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `catch'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
railties (5.0.2) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.0.2) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:69:in `block in tagged'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:69:in `tagged'
railties (5.0.2) lib/rails/rack/logger.rb:24:in `call'
sprockets-rails (3.2.0) lib/sprockets/rails/quiet_assets.rb:13:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/request_id.rb:24:in `call'
rack (2.0.1) lib/rack/method_override.rb:22:in `call'
rack (2.0.1) lib/rack/runtime.rb:22:in `call'
activesupport (5.0.2) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/static.rb:136:in `call'
rack (2.0.1) lib/rack/sendfile.rb:111:in `call'
railties (5.0.2) lib/rails/engine.rb:522:in `call'
puma (3.8.0) lib/puma/configuration.rb:224:in `call'
puma (3.8.0) lib/puma/server.rb:590:in `handle_request'
puma (3.8.0) lib/puma/server.rb:425:in `process_client'
puma (3.8.0) lib/puma/server.rb:289:in `block in run'
puma (3.8.0) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
  Rendering /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout
  Rendering /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (1.3ms)
  Rendered collection of /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/routes/_route.html.erb [1 times] (1.6ms)
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/routes/_table.html.erb (16.9ms)
  Rendering /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (1.5ms)
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (68.0ms)
```

It seems `javascript_pack_tag` is not pointing to the correct location? Am I missing something here?

**Comment by ytbryan** [comment]
Thank you @BigChief45. You forgot about the webpack-dev-server that will serve your assets. 

For production, you can also compile assets using rails webpacker:compile

**Comment by BigChief45** [comment]
@ytbryan Thanks for your reply.

Figured out a few things. I installed Yarn, updated Node, ran `bin/yarn` to update some dependencies, and then re-ran the webpacker and webpacker:vue install commands from before. Everything seems to be alright at this point.

In `development.rb` I set the following configuration:

```ruby
config.x.webpacker[:dev_server_host] = '0.0.0.0'
```

Then I run `bin/webpacker-dev-server`:

```
Project is running at http://localhost:8080/
webpack output is served from /
Content not from webpack is served from /home/ubuntu/workspace/sample/public/packs
(node:2552) DeprecationWarning: loaderUtils.parseQuery() received a non-string value which can be problematic, see https://github.com/webpack/loader-utils/issues/56
parseQuery() will be replaced with getOptions() in the next major version of loader-utils.
Hash: 817ffe00862acdaea322
Version: webpack 2.2.1
Time: 5466ms
             Asset    Size  Chunks                    Chunk Names
      hello_vue.js  654 kB       0  [emitted]  [big]  hello_vue
    application.js  337 kB       1  [emitted]  [big]  application
  hello_vue.js.map  764 kB       0  [emitted]         hello_vue
application.js.map  375 kB       1  [emitted]         application
chunk    {0} hello_vue.js, hello_vue.js.map (hello_vue) 614 kB [entry] [rendered]
   [34] (webpack)-dev-server/client?http://localhost:8080 5.44 kB {0} {1} [built]
   [50] ./~/sockjs-client/lib/entry.js 244 bytes {0} {1} [built]
   [76] ./~/strip-ansi/index.js 161 bytes {0} {1} [built]
   [78] ./~/url/url.js 23.3 kB {0} {1} [built]
   [79] ./~/url/util.js 314 bytes {0} {1} [built]
   [80] (webpack)-dev-server/client/overlay.js 3.6 kB {0} {1} [built]
   [81] (webpack)-dev-server/client/socket.js 856 bytes {0} {1} [built]
   [83] (webpack)/hot/emitter.js 77 bytes {0} {1} [built]
   [85] ./app/javascript/packs/hello_vue.js 551 bytes {0} [built]
   [86] ./~/babel-loader/lib!./~/vue-loader/lib/selector.js?type=script&index=0!./app/javascript/packs/app.vue 114 bytes {0} [built]
   [93] ./app/javascript/packs/app.vue 1.57 kB {0} [built]
   [94] ./~/vue-loader/lib/component-normalizer.js 1.15 kB {0} [built]
   [96] ./~/vue-style-loader!./~/css-loader?sourceMap!./~/vue-loader/lib/style-rewriter.js?{"id":"data-v-55a08a3f","scoped":true,"hasInlineConfig":false}!./~/vue-loader/lib/selector.js?type=styles&index=0!./app/javascript/packs/app.vue 1.57 kB {0} [built]
   [99] ./~/vue/dist/vue.esm.js 244 kB {0} [built]
  [101] multi (webpack)-dev-server/client?http://localhost:8080 ./app/javascript/packs/hello_vue.js 40 bytes {0} [built]
     + 85 hidden modules
chunk    {1} application.js, application.js.map (application) 301 kB [entry] [rendered]
   [34] (webpack)-dev-server/client?http://localhost:8080 5.44 kB {0} {1} [built]
   [35] ./~/ansi-html/index.js 4.26 kB {0} {1} [built]
   [38] ./~/events/events.js 8.33 kB {0} {1} [built]
   [39] ./~/html-entities/index.js 231 bytes {0} {1} [built]
   [44] ./~/punycode/punycode.js 14.7 kB {0} {1} [built]
   [47] ./~/querystring-es3/index.js 127 bytes {0} {1} [built]
   [50] ./~/sockjs-client/lib/entry.js 244 bytes {0} {1} [built]
   [76] ./~/strip-ansi/index.js 161 bytes {0} {1} [built]
   [78] ./~/url/url.js 23.3 kB {0} {1} [built]
   [79] ./~/url/util.js 314 bytes {0} {1} [built]
   [80] (webpack)-dev-server/client/overlay.js 3.6 kB {0} {1} [built]
   [81] (webpack)-dev-server/client/socket.js 856 bytes {0} {1} [built]
   [83] (webpack)/hot/emitter.js 77 bytes {0} {1} [built]
   [84] ./app/javascript/packs/application.js 515 bytes {1} [built]
  [100] multi (webpack)-dev-server/client?http://localhost:8080 ./app/javascript/packs/application.js 40 bytes {1} [built]
     + 71 hidden modules

WARNING in EnvironmentPlugin - NVM_CD_FLAGS environment variable is undefined.

You can pass an object with default values to suppress this warning.
See https://webpack.js.org/plugins/environment-plugin for example.
```

Then in a separate terminal, I run the Rails server on port 8081. When I open in the browser I still get a similar error to before, but this time there is a different path in the error message:

```
Started GET "/javascripts/0.0.0.0/hello_vue.js" for 10.240.0.5 at 2017-03-16 08:38:41 +0000
Cannot render console from 10.240.0.5! Allowed networks: 127.0.0.1, ::1, 127.0.0.0/127.255.255.255
  
ActionController::RoutingError (No route matches [GET] "/javascripts/0.0.0.0/hello_vue.js"):
  
actionpack (5.0.2) lib/action_dispatch/middleware/debug_exceptions.rb:53:in `call'
web-console (3.4.0) lib/web_console/middleware.rb:135:in `call_app'
web-console (3.4.0) lib/web_console/middleware.rb:20:in `block in call'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `catch'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
railties (5.0.2) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.0.2) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:69:in `block in tagged'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.0.2) lib/active_support/tagged_logging.rb:69:in `tagged'
railties (5.0.2) lib/rails/rack/logger.rb:24:in `call'
sprockets-rails (3.2.0) lib/sprockets/rails/quiet_assets.rb:13:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/request_id.rb:24:in `call'
rack (2.0.1) lib/rack/method_override.rb:22:in `call'
rack (2.0.1) lib/rack/runtime.rb:22:in `call'
activesupport (5.0.2) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.2) lib/action_dispatch/middleware/static.rb:136:in `call'
rack (2.0.1) lib/rack/sendfile.rb:111:in `call'
railties (5.0.2) lib/rails/engine.rb:522:in `call'
puma (3.8.0) lib/puma/configuration.rb:224:in `call'
puma (3.8.0) lib/puma/server.rb:590:in `handle_request'
puma (3.8.0) lib/puma/server.rb:425:in `process_client'
puma (3.8.0) lib/puma/server.rb:289:in `block in run'
puma (3.8.0) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
  Rendering /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout
  Rendering /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (7.9ms)
  Rendered collection of /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/routes/_route.html.erb [1 times] (1.6ms)
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/routes/_table.html.erb (1.8ms)
  Rendering /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (8.1ms)
  Rendered /usr/local/rvm/gems/ruby-2.3.0@global/gems/actionpack-5.0.2/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (71.5ms)
```

I feel I'm closer this time, but not sure what else I'm missing here.

**Comment by renchap** [comment]
This is not correct: `config.x.webpacker[:dev_server_host] = '0.0.0.0'`

Try with `config.x.webpacker[:dev_server_host] = 'http://localhost:8080'`

**Comment by BigChief45** [comment]
@renchap Thanks for chiming in. I tried using that configuration, and while I don't get errors anymore, they Vue component is not displaying the "hello" message.

I am working on a cloud based development environment so I think the `localhost` configuration might not work. Any chance on running both Rails and Webpack server in `0.0.0.0` under different ports? I tried checking `app/config/webpack/development.js` for some port option but I can't find any.

**Comment by ytbryan** [comment]
@BigChief45 do you have the repo on github somewhere? Which cloud-based dev environment? Heroku? 😅

Anyway let's take this off this PR request thread. You can open a issue and we can fix it there.

**Comment by renchap** [comment]
@BigChief45 you can configure the host when launching `./bin/webpack-dev-server` : 
```
./bin/webpack-dev-server --host 0.0.0.0
```

Then you need to change `config.x.webpacker[:dev_server_host]` to use your cloud based environment IP or hostname (the same you use to connect to rails). `0.0.0.0` is a special IP meaning "listen on every IP on this computer", you can not use it to connect to it.

If you connect to your Rails app using for example `http://1.2.3.4:3000`, then you need to use `http://1.2.3.4:8080` for `config.x.webpacker[:dev_server_host]`

**Pull request merged**